File: frontend/lib/models/cycle.dart
class CycleLog {
  CycleLog({
    required this.id,
    required this.startDate,
    this.endDate,
    this.cycleLength,
    this.memo,
    this.createdAt,
  });

  final int id;
  final DateTime startDate;
  final DateTime? endDate;
  final int? cycleLength;
  final String? memo;
  final DateTime? createdAt;

  factory CycleLog.fromJson(Map<String, dynamic> json) {
    return CycleLog(
      id: _int(json['id']),
      startDate:
          DateTime.tryParse(json['start_date']?.toString() ?? '') ??
          DateTime.now(),
      endDate: DateTime.tryParse(json['end_date']?.toString() ?? ''),
      cycleLength: json['cycle_length'] == null
          ? null
          : _int(json['cycle_length']),
      memo: json['memo']?.toString(),
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? ''),
    );
  }
}

int _int(dynamic value) => value is int ? value : int.tryParse('$value') ?? 0;

File: backend/app/schemas/cycle.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator


class CycleCreate(BaseModel):
    start_date: date
    end_date: date | None = None
    cycle_length: int | None = Field(default=None, ge=15, le=60)
    memo: str | None = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def validate_date_order(self):
        if self.end_date is not None and self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        return self


class CycleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    start_date: date
    end_date: date | None
    cycle_length: int | None
    memo: str | None
    created_at: datetime
